compiler/plugins/target/LLVMCPU/ResolveCPUAndCPUFeatures.cpp

```diff
@@ -0,0 +1,176 @@
+  
+    cpu="name"
+        Target a specific CPU. This is mostly used on x86. The accepted values are the same as in Clang command lines.)MSG";
+  if (triple.isX86()) {
+    msg += R"MSG(
+        List of accepted x86 CPUs: )MSG";
+    llvm::SmallVector<llvm::StringRef> allAcceptedCpus;
+    llvm::X86::fillValidCPUArchList(allAcceptedCpus, /*Only64Bit=*/true);
+    llvm::raw_string_ostream s(msg);
+    llvm::interleaveComma(allAcceptedCpus, s);
+    msg += "\n";
+  } else {
+    msg += R"MSG(
+        CAVEAT: Outside of x86, this will only set the instruction scheduling model but will not enable CPU features. That's why when targeting non-x86 CPUs, it is usually preferred to pass cpu_features, see below.
+)MSG";
+  }
+  msg += R"MSG(
+    cpu_features="+feature1,..."
+        Target a CPU supporting the comma-separated of (+-prefixed) features. The accepted values are the same as in Clang command lines.
+)MSG";
+  if (triple.isAArch64()) {
+    msg += R"MSG(
+        Example: cpu_features="+dotprod,+i8mm,+bf16
+)MSG";
+  }
+  if (triple.isRISCV()) {
+    msg += R"MSG(
+        Example: cpu_features="+m,+a,+f,+d,+c
+)MSG";
+  }
+  return msg;
+}
+
+} // namespace
+
+ResolveCPUAndCPUFeaturesStatus
+resolveCPUAndCPUFeatures(const llvm::Triple &triple, std::string &cpu,
+                         std::string &cpuFeatures) {
+  // No early-return on error status. The caller may treat these errors as
+  // non-fatal and will carry on with whichever `cpu` and `cpuFeatures` we
+  // produce.
+  auto status1 = cpu.empty() && cpuFeatures.empty()
+                     ? ResolveCPUAndCPUFeaturesStatus::ImplicitGenericFallback
+                     : ResolveCPUAndCPUFeaturesStatus::OK;
+  auto status2 = resolveHostCPUAndCPUFeatures(triple, cpu, cpuFeatures);
+  auto status3 = resolveCPUFeaturesForCPU(triple, cpu, cpuFeatures);
+  tweakCPUFeatures(triple, cpu, cpuFeatures);
+  // Helper to return the first non-OK status.
+  auto combine = [](ResolveCPUAndCPUFeaturesStatus a,
+                    ResolveCPUAndCPUFeaturesStatus b) {
+    return a == ResolveCPUAndCPUFeaturesStatus::OK ? b : a;
+  };
+  return combine(combine(status1, status2), status3);
+}
+
+std::string getMessage(ResolveCPUAndCPUFeaturesStatus status,
+                       const llvm::Triple &triple) {
+  switch (status) {
+  case ResolveCPUAndCPUFeaturesStatus::ImplicitGenericFallback:
+    return getImplicitGenericFallbackMessage(triple);
+  case ResolveCPUAndCPUFeaturesStatus::InconsistentHost:
+    return "If either CPU or CPU-features is `host`, the other must "
+           "be either also `host` or the default value.\n";
+  case ResolveCPUAndCPUFeaturesStatus::UnimplementedMapping:
+    return "Resolution of CPU to CPU-features is not implemented on this "
+           "target architecture. Pass explicit "
+           "CPU-features, or implement the missing mapping.";
```

Comment:
Include `\n` consistently?

```suggestion
    return "If either CPU or CPU-features is `host`, the other must "
           "be either also `host` or the default value.\n";
  case ResolveCPUAndCPUFeaturesStatus::UnimplementedMapping:
    return "Resolution of CPU to CPU-features is not implemented on this "
           "target architecture. Pass explicit "
           "CPU-features, or implement the missing mapping.\n";
```